Add a setting to control applying area transform to gravity

lejar opened the pull request:
Change for #79243

This change adds a new parameter to Area3D, which determines whether or not the gravity calculation of an area should take the area's transform into account.

The default behavior before this setting was to take the transform into account when using the gravity is a point setting, but not when using a simple vector to specify the gravity. The new default value will be true in both cases.

I still need to add some logic to the project upgrade functionality which should disable this setting in older projects for areas using a vector to specify gravity, so that this does not break compatibility, but I couldn't find where this upgrade is done (except the one which upgrades godot 3 -> 4). I would appreciate if someone could point me in the right direction for this.

[test_gravity.zip](https://github.com/godotengine/godot/files/11998155/test_gravity.zip)
Here's a project containing the new setting showing how it works.

* *Bugsquad edit, closes: #79243*
* *Bugsquad edit, alternative to: https://github.com/godotengine/godot/pull/90166*

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

Commit:
Add setting which determines whether or not to apply area transform when calculating gravity.

akien-mga commented:
Thanks for the contribution! The physics team is assigned for review, but it's currently short on contributors, so it might take some time. Feel free to drop by the #physics channel on the [Contributors Chat](https://chat.godotengine.org) to discuss your proposal.

For the record, your commit seems not to be linked to your GitHub account. See: [Why are my commits linked to the wrong user?](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user) for more info.

akien-mga commented on servers/physics_server_3d.h:
```diff
@@ -316,6 +316,7 @@ class PhysicsServer3D : public Object {
 		AREA_PARAM_GRAVITY_OVERRIDE_MODE,
 		AREA_PARAM_GRAVITY,
 		AREA_PARAM_GRAVITY_VECTOR,
+		AREA_PARAM_GRAVITY_APPLY_TRANSFORM,
```
This should be added at the end of the enum to avoid breaking compatibility.

AThousandShips commented:
Should be added to the 2D physics as well for completeness IMO, if added

AThousandShips commented:
Added the fact that this breaks compatibility by the default behaviour of the property, and it's change of behaviour with the point gravity if turned off

AThousandShips commented on servers/physics_3d/godot_area_3d.cpp:
```diff
@@ -327,9 +332,14 @@ void GodotArea3D::call_queries() {
 void GodotArea3D::compute_gravity(const Vector3 &p_position, Vector3 &r_gravity) const {
+	Vector3 gravity_vector = get_gravity_vector();
+	if (gravity_apply_transform) {
+		gravity_vector = get_transform().xform(gravity_vector);
```
The transformation here is different, when a point it should transform the point to local space, but when a direction it should just rotate and scale, not translate, i.e. transform by the basis only

This will make the gravity stronger and stronger the further from origin the area is

AThousandShips replied:
I would say the code for point gravity should be unchanged and the other case should be changed to:
```cpp
	else if (gravity_apply_transform) {
		r_gravity = get_transform().get_basis().xform(get_gravity_vector()) * get_gravity();
	} else {
		r_gravity = get_gravity_vector() * get_gravity();
	}
```

Or similar

lejar replied:
Okay I've done it like you suggested. I agree that this way is better and now this should no longer break compatibility.

AThousandShips reviewed (changes requested):
See above comments on the transformations

Commit:
Move declaration of new enum param to end of enum to avoid breaking compatibility.

Commit:
Make gravity_apply_transform only available / used when gravity is *not* a point. Also change transform to only apply the basis, so that translation is not applied to the gravity vector.

Commit:
Make gravity_apply_transform false by default.

lejar commented:
I've applied the requested changes to the 3D code. I'll work on making the same feature in the 2D code.

Commit:
Do not capitalize non-propper noun in the middle of a sentence.

Commit:
Implement the option for applying area transform to gravity for Area2D.

lejar commented:
Okay I've also gone ahead and implemented the same functionality into Area2D. I've updated my example project to include a 2D scene:
[test_gravity_with_2d.zip](https://github.com/godotengine/godot/files/12024115/test_gravity_with_2d.zip)

AThousandShips commented:
For 2D you instead use `basis_xform` on the `Transform2D`

AThousandShips commented:
The documentation has to be in alphabetical order, you get this automatically if you compile the code and run `--doctool` on the compiled program

AThousandShips commented on doc/classes/Area2D.xml:
```diff
@@ -90,6 +90,9 @@
 			The above is true only when the unit distance is a positive number. When this is set to 0.0, the gravity will be constant regardless of distance.
 		</member>
+		<member name="gravity_apply_transform" type="bool" setter="set_gravity_apply_transform" getter="is_gravity_apply_transform" default="false">
+			If [code]true[/code], gravity is transformed using the area's transform. In other words, rotating the area with this setting active will also rotate the gravity vector.
```
This implies that only rotation will apply, I think a more general wording is better

AThousandShips reviewed (dismissed):
LGTM, but haven't tested, with the default to false, and applying only basis transform, it makes sense to me

(Note, this is mainly on the logic of the implementation, rather than the validity of the feature, I leave that open)

rburing commented:
Correct me if I'm wrong, but if this feature proposal had an associated [Godot Improvement Proposal](https://github.com/godotengine/godot-proposals), then the answer to

> If this enhancement will not be used often, can it be worked around with a few lines of script?

would be yes: store the original vector in a `Vector3` variable on `_ready`, and apply the area transform whenever necessary, i.e. on `_ready` and whenever the area is transformed.

AThousandShips commented:
That is true, reasonable workaround, maybe add it to the issue and we can discuss that part more there

aaronfranke commented on servers/physics_3d/godot_area_3d.cpp:
```diff
@@ -341,6 +346,8 @@ void GodotArea3D::compute_gravity(const Vector3 &p_position, Vector3 &r_gravity)
 		} else {
 			r_gravity = v.normalized() * get_gravity();
 		}
+	} else if (gravity_apply_transform) {
+		r_gravity = get_transform().get_basis().xform(get_gravity_vector()) * get_gravity();
```
With this code, point gravity will always take the transform into account, but directional gravity will only take the transform into account when `gravity_apply_transform` is true. This seems like an odd inconsistency. However, making it consistent would require breaking behavior compatibility, which is what PR #90166 does.

AThousandShips replied:
I'd say there's a valid use case for not applying transform to the direction, much less so for the point, if we *are* to break compatibility I'd argue it would be to make *both* controllable, not make the directional one local
